docs/RationaleSeq.md

```diff
@@ -0,0 +1,100 @@
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. Our
+intention is to allow analysis and optimization of sequential logic without
+having to reason about timing-dependent behavior. This makes it somewhat
+distinct from the SystemVerilog register wherein one has to worry about reset
+behavior and implementation (latch vs. flip-flop).
+
+`Reg` has four operands:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Required iff
+'reset' is present.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [, %resetValue] ] : $type(input)
+```
+
+### Rationale
+
+Several design decisions were made in defining this op. Mostly, they were
+made to simplify it while still providing the common case. Providing support
+for all flavors of registers is an anti-goal of this op. If an omitted
+feature is needed, it can be added or (if not common or precludes
+optimization) another op could be added.
+
+- Logical features:
+  - Inclusion of optional 'reset' signal: This operand makes lowering to an
+  efficient implementation of reset easier. Omission of it would require some
+  (potentially complex) analysis to find the reset mux if required (which it
+  usually is to ensure efficient synthesis).
+  - Inclusion of 'resetValue': if we have a 'reset' signal, we need to
+  include a value.
+  - Omission of 'enable': enables are easily modeled via a multiplexer on the
+  input with one of the mux inputs as the output of the register. This
+  property makes 'enables' easier to detect than reset in the common case.
+
+- Timing / clocking:
+  - Omission of 'negedge' event on 'clock': this is easily modeled by
+  inverting the clock value.
+  - Omission of 'dual edge' event on 'clock': this is not expected to be
+  terribly common.
+  - Omission of edge conditions on 'reset': The reset style shouldn't affect
+  logical correctness. It should, therefore, be determined by a lowering
+  pass.
+
+### Future considerations
+
+- Enable signal: if this proves difficult to detect (or non-performant if we
+do not detect and generate the SystemVerilog correctly), we can build it into
+the reg op.
+- Reset style and clock style: how should we model posedge vs negedge clocks?
+Async vs sync resets? There are some reasonble options here: attributes on
+this op or `clock` and `reset` types which are parameterized with that
+information.
+- Initial value: this register initializes to `x` (in SystemVerilog
+terminology). We will add an `initialValue` attribute if this proves
+insufficient.
```

Comment:
Our currently supported front-end language, firrtl, uses macros and stuff for initial values and not 'x.  To support this front end, we have to have custom initialization.

Reply:
Can you clarify on what is `x` here? In Verilog this could be any of: (1) uninitialized, (2) a one-time random value, (3) a random value every time you read, (4) the user "doesn't care" what the value is and the compiler can choose anything, (5) verification poison, or something else.

Put differently, how would this compose if the target language is VHDL where you have separeate representations for things like the above? I would like to keep the `seq` and `comb` dialects relatively un-opinionated and not wind up repeating the mistakes of Verilog with imprecision around `x`.

One solution would be to make the initial value an actual MLIR value which can support a connection from another dialect, e.g., this could support a connection from a VHDL constant (unititialized, `U`, or don't care, `-`) or a Verilog `x`.

Reply:
I've kinda been resisting having an initial value attribute since I think it'll affect sequential optimizations. I'd rather blanket set it to (1) uninitialized so as to not preclude any optimizations which need to do any number of things which affect the initial value.

I'll change 'x to uninitialized to clarify to start. If we find in the common case an initial value is necessary, we can add it as an optional attribute.